app/code/core/Mage/GiftMessage/Model/Observer.php

```diff
@@ -87,9 +87,24 @@ public function checkoutEventCreateGiftMessage(Varien_Event_Observer $observer)
         $quote = $observer->getEvent()->getQuote();
         /** @var Mage_Sales_Model_Quote $quote */
         if (is_array($giftMessages)) {
-            foreach ($giftMessages as $entityId => $message) {
+            foreach ($giftMessages as $key => $message) {
                 $giftMessage = Mage::getModel('giftmessage/message');
 
+                // Extract the actual entity ID from the prefixed key
+                $entityId = null;
+                if (str_starts_with($key, 'quote_')) {
+                    $entityId = substr($key, 6); // Remove 'quote_' prefix
+                } elseif (str_starts_with($key, 'item_')) {
+                    $entityId = substr($key, 5); // Remove 'item_' prefix
+                } elseif (str_starts_with($key, 'addr_')) {
+                    $entityId = substr($key, 5); // Remove 'addr_' prefix
+                } elseif (str_starts_with($key, 'addritem_')) {
+                    $entityId = substr($key, 9); // Remove 'addritem_' prefix
+                } else {
+                    // Fallback for backward compatibility
+                    $entityId = $key;
+                }
```

Comment:
Since we don't care about which prefix we actually have, how about this?

```suggestion
                // Extract the actual entity ID from the prefixed key
                if (str_contains($key, '_')) {
                    $entityId = substr($key, strrpos($key, '_') + 1);
                } else {
                    // Fallback for backward compatibility
                    $entityId = $key;
                }
```